File: 003,tencent-50/Py/leetcode-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        carry = 0
        l4 = None  # We use this variable to keep track of head of our result LinkedList
        while(True): # Python's makeshift Do while loop -- see if condition below which breaks it
            sumTotal = l1.val + l2.val + carry
            tempNode = ListNode(sumTotal%10) 
            carry = sumTotal//10
            try:
                l3.next = tempNode
                l3 = l3.next
            except(NameError):
                l3 = tempNode
                l4 = l3
            if(l2.next == None or l1.next == None):
                break
            else:
                l1,l2 = l1.next,l2.next
        if(l2.next == None and l1.next == None and carry ==0):
            return l4
        elif(l2.next == None and l1.next == None and carry ==1):
            l3.next = ListNode(1)
            return l4
        l = l1.next if (l2.next == None and l1.next !=None) else l2.next #select list which still has elements remaining
        # Repeat of previous while loop but only single list and carry -- there should be way to optimize this as it looks redundant
        while(True):
            sumTotal = l.val  + carry
            tempNode = ListNode(sumTotal%10)
            carry = sumTotal//10
            l3.next = ListNode(sumTotal%10)
            l3 = l3.next
            if(l.next == None and carry ==0):
                return l4
            elif(l.next == None and carry == 1):
                l3.next = ListNode(1)
                return l4
            else:
                l = l.next

File: 003,tencent-50/Java/leetcode-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
/** 小技巧：
 * 对于链表问题，返回结果为头结点时，通常需要先初始化一个预先指针 pre，该指针的下一个节点指向真正的头结点head。
 * 使用预先指针的目的在于链表初始化时无可用节点值，而且链表构造过程需要指针移动，进而会导致头指针丢失，无法返回结果。
*/
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode pre = new ListNode(0);
        ListNode cur = pre;
        int carry = 0;// 是否进位，1
        
        int x;
        int y;
        
        while(l1 != null || l2 != null){
            if (l1 == null){
                x = 0;
            }else{
                x = l1.val;
            }
            
            if (l2 == null){
                y = 0;
            }else{
                y = l2.val;
            }
            
            int sum = x + y + carry;
            
            carry = sum / 10;
            sum = sum % 10;
            cur.next = new ListNode(sum);
            
            cur = cur.next;
            
            if (l1 != null){
                l1 = l1.next;
            }
            if (l2 != null){
                l2 = l2.next;
            }
            
        }
        if (carry == 1){
            cur.next = new ListNode(carry);
        }
        return pre.next;
    }
}
